src/content/en/updates/posts/2016/03/web-enabled-internet-of-things.markdown

```diff
@@ -0,0 +1,523 @@
+100 reasons and only send updates when the temperature changes by
+at least 1 degree.
+
+{% highlight javascript %}
+TemperatureCharacteristic.prototype.valueChange = function(value) {
+    this._total += value;
+    this._samples++;
+    
+    if (this._samples < NO_SAMPLES)
+        return;
+        
+    var newValue = Math.round(this._total / NO_SAMPLES);
+    
+    this._total = 0;
+    this._samples = 0;
+    
+    if (this._lastValue && Math.abs(this._lastValue - newValue) < 1)
+        return;
+    
+    this._lastValue = newValue;
+    
+    console.log(newValue);
+    var data = new Buffer(8);
+    data.writeDoubleLE(newValue, 0);
+    
+    if (this._onChange)
+      this._onChange(data);
+};
+{% endhighlight %}
+
+Now that was the temperature sensor. The color LED is even more
+simple. Below you find the object as well as the "read" method.
+The characteristic is configured to allow for "read" and "write"
+operations and has a different UUID than the temperature characteristic.
+
+{% highlight javascript %}
+var ColorCharacteristic = function() {
+  bleno.Characteristic.call(this, {
+    uuid: 'fc0b',
+    properties: ['read', 'write'],
+    value: null
+  });
+  this._value = 'ffffff';
+  this._led = null;
+};
+
+util.inherits(ColorCharacteristic, bleno.Characteristic);
+
+ColorCharacteristic.prototype.onReadRequest = function(offset, callback) {
+  var data = new Buffer(this._value);
+  callback(this.RESULT_SUCCESS, data);
+};
+{% endhighlight %}
+
+As I need to control the LED from the object, I add a
+`this._led member` which I use to store the Johnny-Five LED
+object. I also set the color of the LED to its' default
+value (white).
+
+{% highlight javascript %}
+board.on("ready", function() {
+  ...
+  colorCharacteristic._led = led;
+  led.color(colorCharacteristic._value);
+  led.intensity(30);
+  ...
+}
+{% endhighlight %}
+
+The "write" method received a string (just like "read" sends
+a string), which can consist of a CSS color code (ie. CSS names
+like `rebeccapurple` or hex codes like `#ff00bb`). I use a node
+module called parse-color to always get the hex value which is
+what Johnny-Five expects.
+
+{% highlight javascript %}
+ColorCharacteristic.prototype.onWriteRequest = function(data, offset, withoutResponse, callback) {
+  var value = parse(data.toString('utf8')).hex;
+  if (!value) {
+    callback(this.RESULT_SUCCESS);
+    return;
+  }
+    
+  this._value = value;
+  console.log(value);
+
+  if (this._led)
+      this._led.color(this._value);
+  callback(this.RESULT_SUCCESS);
+};
+{% endhighlight %}
+
+All the above will not work if we don't include the bleno module.
+Recall in the notes above that `eddystone-beacon` will not work
+with bleno unless you use the noble version distributed with it?
+Doing that is luckily quite simple:
+
+{% highlight javascript %}
+var bleno = require('eddystone-beacon/node_modules/bleno');
+var util = require('util');
```

Comment:
it is used together with bleno for object inheriting. The bleno examples use it too